### ISSUE TITLE
Bug: [no-base-to-string] False negative for `unknown` type

### ISSUE THREAD

**Issue by OliverJAsh** [typescript-eslint/typescript-eslint#10862]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=PTAEFEA8AcFMGMAusAmpYCcMHsMBpQBzbRUbAO1gDpRBQcgCgBlRDAS3MIAoBvAXwEoA3PXooEAGwCGGWKHgUAzqUgAuUAFdyAa3LYA7uWEgIMBMjSYc%2BIiVC7KNQDLkTFuy6Qh9IA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oCMBDZRLXxdk%2BKkwDm6MACZwYAL4gFQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
// Expected error, got one. ✅
String({});

declare const x: unknown;
// Expected error, got none. ❌
String(x);
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-base-to-string": 2,
  },
};
```

### tsconfig

```jsonc

```

### Expected Result

Expected error.

### Actual Result

No error.

**Comment by kirkwaiblinger** [comment]
+1

Note this came up in discussion before here: https://github.com/typescript-eslint/typescript-eslint/pull/10005#discussion_r1790660220

and [restrict-template-expressions does flag `unknown`-valued expressions](https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=PTAEFEA8AcFMGMAusAmpYCcMHsMBpQBzbRUbAO1gDpRBQcgCgBlRDAS3MIAoBvAXwEoA3PXooEAGwCGGWKHgUAzqUgAuUAFdyAa3LYA7uWEgIMBMjSYc%2BIiVC7KNQDLkTFuy6QhIgAYASbpF4vIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oCMBDZRLXxdk%2BKkwDm6KAHde0JpAA04bHiKlkFanQbM20FGMpk6%2BRAFti8XufoAPYoeTJKHJqgyQ5CyKoC%2BIP5AA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by bradzacher** [comment]
This is actually working by design - the rule purposely ignores any/unknown.

https://github.com/typescript-eslint/typescript-eslint/blob/main/packages%2Feslint-plugin%2Fsrc%2Frules%2Fno-base-to-string.ts#L254-L257

The reason we ignore it is because we cannot reason about the types. Generally we ignore any/unknown in our rules.

I can see that you added tests for this case in https://github.com/typescript-eslint/typescript-eslint/pull/10432 Kirk heh.

**Comment by kirkwaiblinger** [comment]
Hmm, that makes sense for `any` but wouldn't address `unknown` though

**Comment by bradzacher** [comment]
> but wouldn't address unknown though

You _literally_ added a test for `unknown` - the exact case mentioned in OP!!! 

https://github.com/typescript-eslint/typescript-eslint/pull/10432/files#diff-1fe84c853eb3b36ccef7bb68002461fbf568bfb47fe10a9892f1efe36cb2c765R259

**Comment by Josh-Cena** [comment]
`String(unknown)` is very useful in a generic logging context and we can't really reason about its behavior otherwise. What's a good alternative?

**Comment by kirkwaiblinger** [comment]
> `String(unknown)` is very useful in a generic logging context and we can't really reason about its behavior otherwise. What's a good alternative?

Well, that was basically my question here: https://github.com/typescript-eslint/typescript-eslint/pull/10005#discussion_r1771986890

`String(x)` works reasonably well for all primitives but not most objects. `JSON.stringify(x)` works arguably better for some primtiives, but not at all for others, works well for POJOs, but works badly for other objects, regardless of `.toString()`. 

Why is `String(x)` prohibited for `null | undefined | string | number | bigint | symbol | Function | object` (due to the inclusion of `object`) but allowed for `unknown`? I _do_ think that `String(x)` is quite useful but it feels like it's inconsistent for it to be allowed granted the rest of the rule's behavior.

**Comment by kirkwaiblinger** [comment]
> I can see that you added tests for this case in [#10432](https://github.com/typescript-eslint/typescript-eslint/pull/10432) Kirk heh.

Lol that is pretty funny 😀. Granted the context, I'm pretty sure that that was more of a descriptive snapshot of relevant codepaths that I wanted to be sure my changes didn't change, rather than a normative belief on my part 😁

**Comment by OliverJAsh** [comment]
For logging functions I like to use a `Json` type so I can check that the provided value is definitely serializable:

```ts
export type Json = boolean | number | string | null | JsonArray | JsonRecord
export interface JsonRecord {
  readonly [key: string]: Json
}
export interface JsonArray extends ReadonlyArray<Json> {}

export const log = (value: Json) => {
  // …
};
```

**Comment by bradzacher** [comment]
```ts
log({ toJSON() { throw "Oopsies"; } });
```

Isn't JavaScript just the best language?

**Comment by OliverJAsh** [comment]
Haha, yeah, I know it's not bulletproof, but I think it strikes a good balance.

**Comment by JoshuaKGoldberg** [comment]
`String(any)` is also caught by `no-unsafe-argument`. +1 from me on having this rule ignore `any`.

As for `unknown`, I could see why someone would want to flag it - and also why others would want to allow it for logging/other purposes. Proposal: how about an opt-in option for flagging `unknown`?

**Comment by JoshuaKGoldberg** [comment]
Accepting PRs for an opt-in option with a name like `checkUnknown`.
